File: scripts/ip_lookup.py
import sys
import json
import requests

def lookup_ip(ip_address):
    """
    Lookup informasi IP menggunakan ip-api.com
    Free tier: Unlimited (45 requests/minute)
    """
    try:
        url = "http://ip-api.com/json/" + ip_address + "?fields=66846719"
        
        response = requests.get(url, timeout=10)
        
        if response.status_code == 200:
            data = response.json()
        
            if data.get("status") == "fail":
                error = {"error": data.get("message", "Invalid IP or private IP")}
                print(json.dumps(error))
                return 1

            result = {
                "ip": data.get("query", "N/A"),
                "city": data.get("city", "N/A"),
                "region": data.get("regionName", "N/A"),
                "country": data.get("country", "N/A"),
                "country_code": data.get("countryCode", "N/A"),
                "continent": data.get("continent", "N/A"),
                "postal": data.get("zip", "N/A"),
                "latitude": data.get("lat", "N/A"),
                "longitude": data.get("lon", "N/A"),
                "timezone": data.get("timezone", "N/A"),
                "utc_offset": str(data.get("offset", "N/A")),
                "currency": data.get("currency", "N/A"),
                "calling_code": "+" + str(data.get("callingCode", "N/A")) if data.get("callingCode") else "N/A",
                "languages": "N/A",
                "asn": data.get("as", "N/A"),
                "org": data.get("org", "N/A"),
                "isp": data.get("isp", "N/A"),
            }
            
            print(json.dumps(result, indent=2))
            return 0
        else:
            error = {"error": "API returned status code " + str(response.status_code)}
            print(json.dumps(error))
            return 1
            
    except requests.exceptions.Timeout:
        error = {"error": "Request timeout - API tidak merespon"}
        print(json.dumps(error))
        return 1
    except requests.exceptions.RequestException as e:
        error = {"error": "Network error: " + str(e)}
        print(json.dumps(error))
        return 1
    except Exception as e:
        error = {"error": "Unexpected error: " + str(e)}
        print(json.dumps(error))
        return 1

if __name__ == "__main__":
    if len(sys.argv) < 2:
        error = {"error": "Usage: python ip_lookup.py <IP_ADDRESS>"}
        print(json.dumps(error))
        sys.exit(1)
    
    ip = sys.argv[1]
    sys.exit(lookup_ip(ip))

File: commands/iplookup.js
const { SlashCommandBuilder } = require('discord.js');
const { exec } = require('child_process');
const { promisify } = require('util');

const execPromise = promisify(exec);

module.exports = {
    data: new SlashCommandBuilder()
        .setName('iplookup')
        .setDescription('Lookup informasi IP address')
        .addStringOption(option =>
            option.setName('ip')
                .setDescription('IP address yang mau di-lookup')
                .setRequired(true)),
    
    async execute(interaction) {
        const ip = interaction.options.getString('ip');
        
        const ipRegex = /^(\d{1,3}\.){3}\d{1,3}$/;
        if (!ipRegex.test(ip)) {
            return await interaction.reply({
                content: '❌ Format IP address tidak valid! Contoh: `8.8.8.8`',
                ephemeral: true
            });
        }

        await interaction.deferReply();
        
        try {
            const { stdout, stderr } = await execPromise(`python3 scripts/ip_lookup.py ${ip}`);
            
            if (stderr) {
                console.error('Python stderr:', stderr);
            }
    
            const result = JSON.parse(stdout);
            
            if (result.error) {
                return await interaction.editReply({
                    content: `❌ Error: ${result.error}`,
                    ephemeral: true
                });
            }
            
            await interaction.editReply({
                embeds: [{
                    color: 0x3498db,
                    title: '🌐 IP Address Lookup',
                    thumbnail: {
                        url: `https://flagcdn.com/96x72/${result.country_code.toLowerCase()}.png`
                    },
                    fields: [
                        {
                            name: '📍 IP Address',
                            value: `\`${result.ip}\``,
                            inline: true
                        },
                        {
                            name: '🗺️ Location',
                            value: `${result.city}, ${result.region}`,
                            inline: true
                        },
                        {
                            name: '🌍 Country',
                            value: `${result.country} (${result.country_code})`,
                            inline: true
                        },
                        {
                            name: '📮 Postal Code',
                            value: `\`${result.postal}\``,
                            inline: true
                        },
                        {
                            name: '🌏 Continent',
                            value: `\`${result.continent}\``,
                            inline: true
                        },
                        {
                            name: '🕐 Timezone',
                            value: `${result.timezone} (UTC${result.utc_offset})`,
                            inline: true
                        },
                        {
                            name: '📌 Coordinates',
                            value: `\`${result.latitude}, ${result.longitude}\``,
                            inline: true
                        },
                        {
                            name: '💱 Currency',
                            value: `\`${result.currency}\``,
                            inline: true
                        },
                        {
                            name: '📞 Calling Code',
                            value: `\`${result.calling_code}\``,
                            inline: true
                        },
                        {
                            name: '🗣️ Languages',
                            value: `\`${result.languages}\``,
                            inline: false
                        },
                        {
                            name: '🏢 ISP / Organization',
                            value: `\`${result.org}\``,
                            inline: false
                        },
                        {
                            name: '🔢 ASN',
                            value: `\`${result.asn}\``,
                            inline: false
                        }
                    ],
                    footer: {
                        text: '📡 Data from ipapi.co'
                    },
                    timestamp: new Date()
                }]
            });
            
        } catch (error) {
            console.error('Error executing Python script:', error);
            
            await interaction.editReply({
                content: '❌ Terjadi error saat melakukan IP lookup. Pastikan IP address valid!',
                ephemeral: true
            });
        }
    }
};